File: 125_backpack-ii/backpack-ii.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/backpack-ii
@Language: C++
@Datetime: 14-12-08 06:30
*/

class Solution {
public:
    /**
     * @param m: An integer m denotes the size of a backpack
     * @param A & V: Given n items with size A[i] and value V[i]
     * @return: The maximum value
     */
    int backPackII(int m, vector<int> A, vector<int> V) {
        // write your code here
        vector<int> d(m+1);
        for(int i=1;i<=A.size();++i)
            for(int j=m;j>=A[i-1];--j)
                d[j]= max(d[j],d[j-A[i-1]]+V[i-1]);
        return d[m];
    }
};

File: 125_backpack-ii/backpack-ii.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/backpack-ii
@Language: Python
@Datetime: 14-12-15 17:03
'''

class Solution:
    # @param m: An integer m denotes the size of a backpack
    # @param A & V: Given n items with size A[i] and value V[i]
    # @return: The maximum value
    def backPackII(self, m, A, V):
        # write your code here
        d=[0 for i in xrange(m+1)]
        for i in xrange(1,len(A)+1):
            for j in reversed(xrange(A[i-1],m+1)):
                d[j]=max(d[j],V[i-1]+d[j-A[i-1]])
        return d[m]
